### DISCUSSION TITLE
Generate extension creates 999 sitemap XML files

### DISCUSSION THREAD

**Discussion by JetGum** [datenstrom/community#974] [Report a bug] [answered]
Dear Anna, 

thanks for the Generate extension. I use it for a couple of websites, and I am switching completely to static sites. A great experience.

Still, there is one site that is causing me to scratch my head. 

I have the latest version of Yellow Core 0.9.9 on this particular site, all extensions are up-to-date.
Sitemap 0.9.2 extension is installed, with out-of-the-box settings, and Generate 0.9.5.

So what happens is this:

After the command line (`php yellow.php generate`), the process starts (`Generating static webite`). 
Once it reaches 90%, the process seems to pause for 3-5 minutes. 
The log file does not record any errors. No error messages are displayed.

What happens during the "pause" is this: 

In the /public/sitemap/ folder, the Generate extension creates:
- an index.html for the static site sitemap file,
- a `pagesitemap.xml` file for the Googlebot (yes, that is the filename. The ":" between page and sitemap does not really show).
- 999 single files with the same content as index.html, only with a different and incremental filename. See Screenshot..

![image](https://github.com/datenstrom/yellow/assets/48367334/f4e221f4-985d-4614-8884-b287000aeffb)

`Generating static website 100%... done
Yellow generate: 1341 files, 0 errors`

Is there an explanation and maybe a fix for this?

Cheers,

Carsten

**Comment by markseuffert** [comment]
Does this also happen in a fresh installation plus sitemap? Do you know how to reproduce the problem?

**Reply by JetGum** [reply]
This does not happen in a fresh installation. So the problem seems to be limited to just one of my installations.

**Reply by markseuffert** [reply]
I suggest to investigate a little bit and find out what causes the problem in one of your installations. Perhaps there's a specific setting or Markdown file that triggers the bug? Can you activate the debug mode with `CoreDebugMode: 2` and run `php yellow.php generate | grep sitemap`?

The following is the output in a fresh installation on Linux. This is how it should look like without bug:

```
YellowLookup::findContentLocationFromFile /sitemap/ <- content/sitemap/page.md<br/>
YellowContent::scanLocation location:/sitemap/<br/>
Generating static website 0%... YellowGenerate::analyseLocations detected location:/sitemap/page:sitemap.xml<br/>
Generating static website 50%... YellowGenerate::generateStaticFile status:200 location:/sitemap/<br/>
Generating static website 80%... YellowGenerate::generateStaticFile status:200 location:/sitemap/page:sitemap.xml<br/>
YellowGenerate::generateStaticFile status:100 location:/sitemap/page:sitemap.xml2<br/>
```

**Reply by JetGum** [reply]
I proceeded as suggested, and discovered I made some silly mistakes in my sitemap.html. 
I went back to the original sitemap.html and everything works fine now.

Thanks for your support!

**Reply by markseuffert** [reply]
I would like to understand how to trigger the problem, maybe have an example?

**Reply by JetGum** [reply]
Sure thing. Here is the faulty sitemap.html.

`<?php $this->yellow->layout("header") ?><main><h1><?php echo $this->yellow->page->getHtml("titleContent") ?></h1><ul><?php $section = $sectionNew = "" ?><?php foreach ($this->yellow->page->getPages("sitemap") as $page): ?><?php $sectionNew = htmlspecialchars(strtoupperu(substru($page->get("title"), 0, 1))) ?><?php if ($section!=$sectionNew) { $section = $sectionNew; echo "</ul><h2>$section</h2><ul>\n"; } ?><li><a href="<?php echo $page->getLocation(true) ?>"><?php echo $page->getHtml("titleContent") ?></a></li><?php endforeach ?></ul><?php $this->yellow->layout("pagination", $this->yellow->page->getPages("sitemap")) ?></main><?php $this->yellow->layout("footer") ?>`

**Reply by markseuffert** [reply]
Ah now I understand it. The page collection is never paginated, the API never sets status code 404 and the probing never stops.

**Comment by markseuffert** [comment, marked as answer]
Thank you for the bug report. I changed how locations with arguments are handled and updated the static generator. Here are some technical explanations in case we come back to this bug later. The root cause was not a faulty layout file, but rather how we handled location arguments internally. The static generator has to generate an entire website in advance. It has to analyse and detect all pages that have tags and paginations. As far as I can see in the code base history, a defect has been introduced over two years ago. The pagination has been moved from the extension files to layout files. What should have been a mandatory exit condition has become optional. Summa summarum, the mentioned `sitemap.html` is absolutely fine and the static generator has been updated.

Should be fixed now, please try it out.

**Reply by markseuffert** [reply]
Can someone be so kind and check if this fixes the reported problem?

**Reply by schulle4u** [reply]
I've tested with the original sitemap layout as well as with the `sitemap.html` provided by @JetGum. Works as expected in a fresh installation. 👍
